File: app/lib/services/auth_api.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../models/auth_failure.dart';
import '../models/user.dart';
import 'api_client.dart';

class AuthApi {
  AuthApi(this._dio);
  final Dio _dio;

  Future<({User user, String accessToken, String refreshToken})> login(
    String email,
    String password,
  ) async {
    try {
      final response = await _dio.post('/api/auth/login', data: {
        'email': email,
        'password': password,
      });
      final data = response.data as Map<String, dynamic>;
      return (
        user: User.fromJson(data['user'] as Map<String, dynamic>),
        accessToken: data['accessToken'] as String,
        refreshToken: data['refreshToken'] as String,
      );
    } on DioException catch (e) {
      throw _mapError(e);
    }
  }

  /// Used by AuthNotifier.tryAutoLogin(); the interceptor's own refresh
  /// call is made directly on `dio` in api_client.dart, not through here,
  /// to avoid a needless extra provider hop inside the interceptor closure.
  Future<String> refresh(String refreshToken) async {
    try {
      final response = await _dio.post('/api/auth/refresh', data: {
        'refreshToken': refreshToken,
      });
      return (response.data as Map<String, dynamic>)['accessToken'] as String;
    } on DioException catch (e) {
      throw _mapError(e);
    }
  }

  Future<void> logout(String refreshToken) async {
    try {
      await _dio.post('/api/auth/logout', data: {'refreshToken': refreshToken});
    } on DioException {
      // Best-effort — the API always returns 200 per its contract; ignore failures.
    }
  }

  AuthFailure _mapError(DioException e) {
    final statusCode = e.response?.statusCode;
    if (statusCode == 401) return const AuthFailure.invalidCredentials();
    if (statusCode == 400) {
      final message = (e.response?.data as Map?)?['error'] as String? ?? 'Validation error';
      return AuthFailure.validation(message);
    }
    if (statusCode == 500) return const AuthFailure.server();
    return const AuthFailure.network();
  }
}

final authApiProvider = Provider<AuthApi>((ref) => AuthApi(ref.watch(dioProvider)));

File: app/app/services/authApi.ts
import type { AuthFailure } from '../types/auth'

interface FetchErrorLike {
  response?: {
    status?: number
    _data?: unknown
  }
}

function baseUrl(): string {
  return useRuntimeConfig().public.apiBaseUrl
}

// simplejwt 401s: {detail: "..."}. DRF 400s: per-field array shape, e.g.
// {email: ["This field is required."]} — first field's first message wins.
function extractValidationMessage(data: unknown): string | undefined {
  if (!data || typeof data !== 'object') return undefined
  const obj = data as Record<string, unknown>
  if (typeof obj.detail === 'string') return obj.detail
  for (const value of Object.values(obj)) {
    if (Array.isArray(value) && typeof value[0] === 'string') return value[0]
  }
  return undefined
}

function mapError(error: unknown): AuthFailure {
  const status = (error as FetchErrorLike).response?.status
  if (status === 401) return { kind: 'invalid-credentials' }
  if (status === 400) {
    const message = extractValidationMessage((error as FetchErrorLike).response?._data) ?? 'Validation error'
    return { kind: 'validation', message }
  }
  if (status === 500) return { kind: 'server' }
  return { kind: 'network' }
}

export const authApi = {
  async login(email: string, password: string) {
    try {
      return await $fetch<{ access: string; refresh: string }>('/api/token/', {
        baseURL: baseUrl(),
        method: 'POST',
        body: { email, password },
      })
    } catch (error) {
      throw mapError(error)
    }
  },

  // Called directly by AuthStore.tryAutoLogin(); apiClient's own 401
  // interceptor makes its refresh call inline rather than through here, to
  // avoid an extra module hop inside that hot path.
  async refresh(refreshToken: string): Promise<string> {
    try {
      const data = await $fetch<{ access: string }>('/api/token/refresh/', {
        baseURL: baseUrl(),
        method: 'POST',
        body: { refresh: refreshToken },
      })
      return data.access
    } catch (error) {
      throw mapError(error)
    }
  },

  async logout(refreshToken: string): Promise<void> {
    try {
      await $fetch('/api/token/blacklist/', {
        baseURL: baseUrl(),
        method: 'POST',
        body: { refresh: refreshToken },
      })
    } catch {
      // Best-effort — always treat logout as succeeding client-side.
    }
  },
}
